v2/test/e2e/mpi_job_test.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2021 The Kubeflow Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"context"
+
+	common "github.com/kubeflow/common/pkg/apis/common/v1"
+	"github.com/onsi/ginkgo"
+	"github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	kubeflow "github.com/kubeflow/mpi-operator/v2/pkg/apis/kubeflow/v2beta1"
+)
+
+var _ = ginkgo.Describe("MPIJob", func() {
+	var (
+		namespace string
+		mpiJob    *kubeflow.MPIJob
+	)
+
+	ginkgo.BeforeEach(func() {
+		ns := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: "e2e-",
+			},
+		}
+		var err error
+		ns, err = k8sClient.CoreV1().Namespaces().Create(context.Background(), ns, metav1.CreateOptions{})
+		gomega.Expect(err).ToNot(gomega.HaveOccurred())
+		namespace = ns.Name
+	})
+
+	ginkgo.AfterEach(func() {
+		if namespace != "" {
+			err := k8sClient.CoreV1().Namespaces().Delete(context.Background(), namespace, metav1.DeleteOptions{})
+			if !errors.IsNotFound(err) {
+				gomega.Expect(err).ToNot(gomega.HaveOccurred())
+			}
+		}
+	})
+
+	ginkgo.BeforeEach(func() {
```

Comment:
It's maybe more philosophical, but I think that all logic specific to openMPI should be encapsulated within "ginkgo.Context("with OpenMPI implementation", func() { ... }). Is the mpiJob's spec identical for intelMPI apart from the image? Would it work to have a function "func getPiMPIJob(image string)" and use it as "getPiMPIJob(openMPIImage)"? If not, maybe just move it to the OpenMPI context?

Reply:
The command changes a little too. Moved to Context